target/linux/generic/files/drivers/platform/mikrotik/mikrotik_wlan_lz77.c

```diff
@@ -0,0 +1,443 @@
+ * @in_offset_bit:	bit offset where count starts
+ * @shift:		left shift operand value of first count bit
+ * @count:		initial count
+ * @bits_used:		how many bits were consumed by this count
+ * @max_bits:		maximum bit count for this counter
+ *
+ * Returns the decoded count
+ */
+static int mikrotik_wlan_lz77_decode_count(
+		const u8 *in,
+		size_t in_offset_bit,
+		u8 shift,
+		size_t count,
+		u8 *bits_used,
+		u8 max_bits)
+{
+	size_t pos = in_offset_bit;
+	size_t max_pos = pos + max_bits;
+	bool up = true;
+
+	*bits_used = 0;
+	pr_debug(MIKRO_LZ77 "decode_count inbit: %zu, start shift:%u, initial count:%zu\n",
+			in_offset_bit, shift, count);
+
+	while (true) {
+		if (pos >= max_pos) {
+			pr_err(MIKRO_LZ77 "max bit index reached before count completed\n");
+			return -EINVAL;
+		}
+
+		/* if the bit value at offset is set */
+		if (mikrotik_wlan_lz77_get_bit(in, pos))
+			count += (1 << shift);
+
+		/* shift increases until we find an unsed bit */
+		else if (up)
+			up = false;
+
+		if (up)
+			++shift;
+		else {
+			if (!shift) {
+				*bits_used = pos - in_offset_bit + 1;
+				return count;
+			}
+			--shift;
+		}
+
+		++pos;
+	}
+
+	return -EINVAL;
+}
+
+/**
+ * mikrotik_wlan_lz77_decode_instruction
+ *
+ * @in:			compressed data
+ * @in_offset_bit:	bit offset where instruction starts
+ * @bits_used:		how many bits were consumed by this count
+ *
+ * Returns the decoded instruction
+ */
+static enum mikrotik_wlan_lz77_instruction mikrotik_wlan_lz77_decode_instruction(
+		const u8 *in,
+		size_t in_offset_bit,
+		u8 *bits_used)
+{
+	if (mikrotik_wlan_lz77_get_bit(in, in_offset_bit)) {
+		*bits_used = 2;
+		if (mikrotik_wlan_lz77_get_bit(in, ++in_offset_bit))
+			return INSTR_LONG;
+		else
+			return INSTR_PREVIOUS_OFFSET;
+	} else {
+		*bits_used = 1;
+		return INSTR_LITERAL_BYTE;
+	}
+	return INSTR_ERROR;
+}
+
+/**
+ * mikrotik_wlan_lz77_decode_instruction_operators
+ *
+ * @in:			compressed data
+ * @in_offset_bit:	bit offset where instruction starts
+ * @previous_offset:	last used match offset
+ * @opcode:		struct to hold instruction & operators
+ *
+ * Returns error code
+ */
+static int mikrotik_wlan_lz77_decode_instruction_operators(
+		const u8 *in,
+		size_t in_offset_bit,
+		size_t previous_offset,
+		struct lz77_mk_instr_opcodes *opcode)
+{
+	enum mikrotik_wlan_lz77_instruction instruction;
+	u8 bit_count = 0;
+	u8 bits_used = 0;
+	int offset = 0;
+	int length = 0;
+
+	instruction = mikrotik_wlan_lz77_decode_instruction(
+			in, in_offset_bit, &bit_count);
+
+	/* skip bits used by instruction */
+	bits_used += bit_count;
+
+	switch (instruction) {
+	case INSTR_LITERAL_BYTE:
+		/* non-matching char */
+		offset = 0;
+		length = 1;
+		break;
+
+	case INSTR_PREVIOUS_OFFSET:
+		/* matching group uses previous offset */
+		offset = previous_offset;
+
+		length = mikrotik_wlan_lz77_decode_count(
+				in,
+				in_offset_bit + bits_used,
+				0, 1, &bit_count,
+				MIKRO_LZ77_MAX_COUNT_BIT_LEN);
+		if (length < 0)
```

Comment:
`if (unlikely(length < 0))`
Same below and throughout the rest of the code: wrap all your sanity checks into `unlikely()` (and your most likely cases into `likely()`) to make sure the compiler optimizes for the fast path.

Reply:
thanks, added these optimisation hints throughout.